types/decimal.go

```diff
@@ -0,0 +1,356 @@
+package types
+
+import (
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"strconv"
+	"strings"
+	"testing"
+)
+
+// NOTE: never use new(Dec) or else we will panic unmarshalling into the
+// nil embedded big.Int
+type Dec struct {
+	*big.Int `json:"int"`
+}
+
+// number of decimal places
+const Precision = 10
+
+var (
+	precisionExpReuse    = big.NewInt(Precision)
+	ten                  = big.NewInt(10)
+	precisionReuse       = new(big.Int).Exp(big.NewInt(10), big.NewInt(Precision), nil)
+	precisionMultipliers []*big.Int
+)
+
+// Set precision multipliers
+func init() {
+	precisionMultipliers = make([]*big.Int, Precision+1)
+	for i := 0; i <= Precision; i++ {
+		precisionMultipliers[i] = calcPrecisionMultiplier(int64(i))
+	}
+}
+
+func precisionInt() *big.Int {
+	return new(big.Int).Set(precisionReuse)
+}
+
+// nolint - common values
+func ZeroDec() Dec { return Dec{big.NewInt(0)} }
```

Comment:
Do we need this to be a function - why not a constant (since all the mutating functions copy anyways)? Would save some allocation.

Reply:
An external caller could do `ZeroDec().Int.Add(one, one)`.

However this should cache the big.NewInt(0), and just call a copy.

Reply:
addressed